### PR TITLE
[PkgConfig] Consider quotes when parsing pc file key-value pairs

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
- https://bugs.swift.org/browse/SR-2898

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

File: Sources/Utility/PkgConfig.swift [hunk 1]
```diff
@@ -282,12 +282,16 @@ struct PkgConfigParser {
         }
         
         var it = line.characters.makeIterator()
+        // Indicates if we're in a quoted fragment, we shouldn't append quote.
+        var inQuotes = false
         while let char = it.next() {
-            if char == "\\" {
+            if char == "\"" {
```
Comment (ddunbar) [inline Sources/Utility/PkgConfig.swift]: What does pkg-config do one strings like: `hello"small"world"`?

File: Sources/Utility/PkgConfig.swift [hunk 2]
```diff
@@ -282,12 +282,16 @@ struct PkgConfigParser {
...
+                inQuotes = !inQuotes
+            } else if char == "\\" {
                 if let next = it.next() {
                     fragment.append(next)
                 }
-            } else if char == " " {
+            } else if char == " " && !inQuotes {
                 saveFragment()
```
Comment (ddunbar) [inline Sources/Utility/PkgConfig.swift]: Shouldn't we remove the quotes from a quoted fragment?

Comment (aciidb0mb3r) [inline Sources/Utility/PkgConfig.swift]: yes this removes the quote or rather doesn't add them at all to the fragment

Comment (ddunbar) [inline Sources/Utility/PkgConfig.swift]: oh, ya, right!

File: Sources/Utility/PkgConfig.swift (hunk 1 above)
Comment (aciidb0mb3r) [inline Sources/Utility/PkgConfig.swift]: Couldn't parse Libs field into an argument vector: Text ended before matching quote was found for ". (The text was '-L"/usr/local/bin" -l"gt"k 3" -wantareal\\here -one\\ -two')

Comment (ddunbar) [inline Sources/Utility/PkgConfig.swift]: Ok, cool. Can we also throw an error on those cases then?

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

File: Sources/Utility/PkgConfig.swift [hunk 3]
```diff
@@ -282,17 +282,24 @@ struct PkgConfigParser {
...
             } else {
                 fragment.append(char)
             }
         }
+        guard !inQuotes else {
+            throw PkgConfigError.parsingError("Text ended before matching quote was found in line: \(line) file: \(pcFile.asString)")
```
Comment (ddunbar) [inline Sources/Utility/PkgConfig.swift]: This is a good check, but we also are doing something different than pkg-config on cases where the quotes are inside the argument, right?

Comment (aciidb0mb3r) [inline Sources/Utility/PkgConfig.swift]: different how?

Comment (aciidb0mb3r) [inline Sources/Utility/PkgConfig.swift]: pkgconfig removes the quotes and after this patch our parser will too

**Pull request merged**